docs/influence-sensor-execution-order.md

```diff
@@ -0,0 +1,73 @@
+# Influencing the Sensors execution order
+
+## Preamble
+
+The following document applies for 3 kinds of classes in plugins, the ones implementing:
+* `org.sonar.api.batch.sensor.Sensor`
+* `org.sonar.api.scanner.sensor.ProjectSensor`
+* `org.sonar.api.batch.postjob.PostJob`
+
+For the rest of the document we will refer to them using the generic term "Sensors", except if explicitly stated otherwise.
+
+## Introduction
+
+In Sonar products, Sensors are executed one after another. In some situations, it can be useful for a Sensor to always be executed after another one; one use case is that one Sensor builds an intermediate representation of the code, and other Sensors can compute issues from it. There are different mechanisms in the API for plugins to influence the execution order.
+
+## Registering for an execution phase
+
+Sensors are executed by products during a `Phase`. There are 3 of them:
+
+* PRE
+* DEFAULT
+* POST
+
+The phases execute in this order: PRE -> DEFAULT -> POST.
+
+Each Sensor can register for a specific phase. By default, they belong to the `DEFAULT` phase. This can be changed by annotating the Sensor class with the `Phase` annotation, as follows:
+
+```
+import org.sonar.api.batch.Phase;
+import org.sonar.api.batch.sensor.Sensor;
+
+
+@Phase(name = Phase.Name.PRE)
+public class MySensor implements Sensor {
+
+}
+```
+
+As Phases are executed in a predefined order, registering Sensors in different phases enforces their order of execution.
+
+## Ordering Sensors within a phase
+
+Within those phases, by default Sensors are executed in a random order. It means that from one execution to another, the order is not guaranteed.
+
+There is a mechanism to enforce Sensors to be executed in a defined order, by explicitly marking the dependency between Sensors. To achieve that, the two `@DependsUpon` and `@DependedUpon` annotations can be used:
+
+
+```
+import org.sonar.api.batch.DependedUpon;
+import org.sonar.api.batch.DependsUpon;
+import org.sonar.api.batch.sensor.Sensor;
+
+
+@DependedUpon(value = "UniqueKey")
+public class SensorToBeExecutedFirst implements Sensor {
+
+}
+
+@DependsUpon(value = "UniqueKey")
+public class SensorToBeExecutedSecond implements Sensor {
+
+}
+```
+
+The `DependedUpon` and `DependsUpon` annotations serve to enforce a dependency between sensors. The execution order is guaranteed: Sensors annotated with `DependedUpon` will be executed before Sensors annotated with `DependsUpon`, for Sensors who share the same dependency `value`.
+
```

Comment:
Using a `String` as a dependency key allows a loose coupling. 
The scanner will create a DAG (directed acyclic graph) when resolving dependencies. If there is any cycle, the execution will fail.
The dependency relation is optional: If a Sensor declares a `@DependsUpon(value = "UniqueKey")` and no other Sensor has the matching `@DependedUpon(value = "UniqueKey")`, then the Sensor will be executed in random order.

If multiple Sensors S1, S2, S3 have the same `@DependedUpon(value = "UniqueKey")` and multiple Sensors S4, S5, S6 have the same `@DependsUpon(value = "UniqueKey")`, then the only guarantee is that S4, S5 and S6 will be executed after S1, S2 and S3.

Be aware that a Sensor having a `@DependedUpon(value = "UniqueKey")` is opening the door for other Sensors to depends on it. This is an "API". Later removing the annotation or changing the key can be a breaking change.

Reply:
It's interesting, thanks. I think there are a few things here that are implementation details, so I won't mention them. There are other things that should be mentioned, like the last paragraph.